### PR TITLE
fix(dynamic): replace semicolons with periods in tool descriptions

### PR DESCRIPTION
**Pull request by B1tem3250**
Some MCP gateway implementations validate the tool descriptions returned by tools/list and reject characters they treat as unsafe, including the semicolon. When such a gateway introspects this server it fails with an error like:

    All tools failed validation. First error:
    gitlab_execute_action: Description contains unsafe characters: ';'

This blocks onboarding the server behind those gateways even though the semicolons are ordinary English punctuation in the dynamic toolset's descriptions.

Replace the semicolons with periods in the dynamic surface descriptions (gitlab_find_action, gitlab_execute_action), the execute-envelope hint, the confirm parameter schema, and the destructive-action x_confirmation note in the resource tool manifest. The wording and meaning are preserved; only punctuation changes.

Regenerate the LLM catalog files (llms*.txt) and the LHM plugin manifest so the freshness gates stay green, and relax the case-sensitive "read-only" assertion in the dynamic register test since the phrase now starts a sentence.

## Related Issue

Closes #

## Type of Change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Enhancement (improvement to existing functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring (no functional change)
- [ ] Documentation update
- [ ] CI / build / tooling

## Changes Made

- Replaced semicolons with periods in the dynamic surface tool descriptions (`gitlab_find_action`, `gitlab_execute_action`), the execute-envelope hint, and the `confirm` parameter `jsonschema` tag in `internal/tools/dynamic/register.go`
- Replaced the semicolon in the destructive-action `x_confirmation` description in `internal/resources/tool_manifest.go`
- Relaxed the case-sensitive `"read-only"` assertion in `internal/tools/dynamic/register_test.go` (the phrase now begins a sentence, so it is capitalised)
- Regenerated `llms.txt`, `llms-full.txt`, `llms-medium.txt`, `llms-full-meta-tools.txt`, and `lhm.plugin.json` via `cmd/gen_llms` and `cmd/gen_lhm_manifest`

## How to Test

1. Run `go run ./cmd/gen_llms/ --check` and `go run ./cmd/gen_lhm_manifest/ --check` — both report current. `grep "action catalog" llms.txt` shows periods, not semicolons.
2. Run `go test ./internal/tools/ ./internal/tools/dynamic/ ./internal/resources/ -count=1` — all pass, including `TestToolSnapshots`.
3. Register the server behind an MCP gateway that validates tool descriptions (verified against IBM ContextForge / mcp-context-forge). Before this change onboarding failed with `Description contains unsafe characters: ';'`. After, the gateway onboards the server, lists both tools, and routes a live `project.get` call through to a self-managed GitLab instance.

## Breaking Changes / Migration Notes

N/A

### Code Quality

- [ ] Code compiles: `go build ./...`
- [ ] Consolidated Go analysis passes on changed packages (or `make golangci-lint` / `make analyze`)
- [ ] Code is formatted: `make analyze-fix` (runs configured `golangci-lint` formatters: `goimports`, `gofumpt`, and `gci`)
- [ ] Follows idiomatic Go patterns and the conventions documented in `CLAUDE.md` / `.github/instructions/`

### Testing

- [ ] All existing tests pass: `go test ./... -count=1`
- [ ] New tests added for new functionality (table-driven, with `httptest` mocks)
- [ ] Coverage on modified packages is ≥ 90% (per project policy)
- [ ] Edge cases and error scenarios covered
- [ ] If applicable, E2E tests updated/added under `test/e2e/suite/`

### Documentation

- [ ] Doc comments added for exported types/functions (godoc)
- [ ] `docs/` updated if public API or behavior changed
- [ ] `README.md` / Starlight site (`site/src/content/docs/`) updated if user-facing behavior changed
- [ ] If introducing/removing a tool: `docs/tools/{domain}.md` and tool counts updated
- [ ] Commit messages follow [Conventional Commits](https://www.conventionalcommits.org/)

### Security

- [ ] No secrets, tokens, or credentials in code, tests, fixtures, or logs
- [ ] Input validation for user-provided parameters
- [ ] Error messages do not leak sensitive information
- [ ] No new dependencies with known vulnerabilities (verify with `govulncheck`)

## Screenshots / Logs (if applicable)

<a href="https://cubic.dev/pr/jmrplens/gitlab-mcp-server/pull/341?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

- **Documentation**
  - Clarified descriptions for GitLab action discovery and execution.
  - Improved guidance for destructive actions, including placing `confirm=true` at the top level rather than inside parameters.
  - Reformatted tool documentation into clearer, easier-to-read sentences.
  - Updated descriptions consistently across generated tool metadata and reference documentation.

- **Style**
  - Improved wording, punctuation, and readability without changing functionality or behavior.

### CODE REVIEW

File: llms.txt [hunk 1]
```diff
@@ -61,7 +61,7 @@ Dynamic toolset (default mode):
 
 When TOOL_SURFACE is unset or set to dynamic, the server exposes only gitlab_find_action and gitlab_execute_action while keeping the same canonical GitLab action catalog. Models should find an action with its exact schema, then execute the canonical domain.action ID returned by find. Set TOOL_SURFACE=meta to use consolidated domain meta-tools instead.
 
-- gitlab_find_action: Search the local GitLab action catalog; read-only and no GitLab API call.
+- gitlab_find_action: Search the local GitLab action catalog.
```
[the remark this answers is not in the document]
Comment (B1tem3250) [reply llms.txt:64]: This is expected behaviour, not a regression. `llms.txt` is the compact catalog — `cmd/gen_llms` intentionally reduces every tool line to its first sentence via `firstSentence()` (see `cmd/gen_llms/main.go:308` and the doc comment at line 1054). Every other tool is truncated the same way, e.g. `gitlab_ci_variable`, `gitlab_custom_emoji`.

Previously the `gitlab_find_action` description was a single sentence joined by a semicolon, so the whole thing survived truncation. Now that the semicolon is a period, `firstSentence()` correctly stops at the first sentence — consistent with the rest of the file.

The complete tool contract is preserved in `llms-full.txt`:

> Search the local GitLab action catalog. Read-only and no GitLab API call. Use when the action ID or params are unclear. Returns schemas, hints, destructive flags, and execute examples.

And the runtime metadata in `register.go` is unchanged in meaning. So clients get the full description at runtime and via `llms-full.txt`; `llms.txt` stays a first-sentence summary by design.

**Comment by jmrplens** [comment]
Thanks for this, and for pinpointing the gateway and the exact rejection. I investigated on both sides before deciding how to land it.

On the gateway side: IBM mcp-context-forge removed the semicolon from its forbidden description patterns in 0.7.0 (https://github.com/IBM/mcp-context-forge/issues/3770, fixed by https://github.com/IBM/mcp-context-forge/pull/3916), so upgrading the gateway also resolves this exact rejection; on older versions `TOOL_DESCRIPTION_FORBIDDEN_PATTERNS` tunes the rule and `VALIDATION_STRICT=false` downgrades it to a warning.

On the server side your diff is right in spirit but covers one of the three tool surfaces: the same semicolons were served by the meta and individual surfaces, the prompt and resource listings, and the descriptions embedded in tool schemas, and most of them came from generated format strings rather than literals. I have opened https://github.com/jmrplens/gitlab-mcp-server/pull/342, which makes every served surface semicolon-free, adds a CI audit so it stays that way, and adds a `GITLAB_MCP_DESCRIPTION_SUBSTITUTIONS` knob so a deployment can meet whatever character rule the next gateway enforces without waiting for a release. That PR supersedes this one and credits you as co-author of the cleanup, so I will close this PR when it merges. Thanks again for the report, it is what got the whole surface cleaned.

**Comment by B1tem3250** [comment]
Thanks for the thorough response and for taking it further than my patch did.

Good to know 0.7.0 already dropped the semicolon from the forbidden patterns
(IBM/mcp-context-forge#3770 / #3916) — I'll flag the gateway upgrade path and the
`TOOL_DESCRIPTION_FORBIDDEN_PATTERNS` / `VALIDATION_STRICT` knobs to our platform
team as the cleaner fix on our side.

#342 is clearly the right call — covering the meta and individual surfaces, the
prompt/resource listings, and the schema-embedded descriptions, plus the CI audit
and the `GITLAB_MCP_DESCRIPTION_SUBSTITUTIONS` knob, addresses the whole class of
problem rather than the two dynamic-surface descriptions I caught. The
substitutions env var in particular is a nice touch for the "next gateway with a
different rule" case.

Happy for this to be superseded — please go ahead and close it once #342 merges.
Thanks for the co-author credit, and for the quick turnaround.

**Comment by jmrplens** [comment]
Superseded: the gateway-compat layer (PR 342, with your co-authorship on the substitution commit) merged into main today with the rest of the stack, so the semicolons are gone from every served description and the substitution knob covers whatever rule a gateway adds next. Thank you again for the report and the patch.
